File: server/src/bootstrap.ts
import { prisma } from "./db.js";
import { hashPassword } from "./auth/hash.js";
import { normalizeUsername } from "./username.js";

/**
 * Makes sure the first admin account exists on startup, so a fresh
 * deployment is usable without shell access to run a seed script.
 * Idempotent: an existing account is left untouched (password included).
 */
export async function ensureAdminExists(): Promise<void> {
  const email = process.env.ADMIN_EMAIL;
  const username = process.env.ADMIN_USERNAME;
  const password = process.env.ADMIN_PASSWORD;

  if (!email || !username || !password) return;

  const existing = await prisma.user.findFirst({
    where: { OR: [{ email }, { usernameLower: normalizeUsername(username) }] },
  });
  if (existing) return;

  await prisma.user.create({
    data: {
      firstName: process.env.ADMIN_FIRST_NAME || "Admin",
      lastName: process.env.ADMIN_LAST_NAME || "Admin",
      email,
      username,
      usernameLower: normalizeUsername(username),
      passwordHash: await hashPassword(password),
      role: "ADMIN",
      status: "ACTIVE",
      canEditProfile: true,
    },
  });
  console.log(`Created initial admin account: ${username}`);
}

File: backend/app/services/bootstrap.py
import os

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.hash import hash_password
from app.models.enums import Role, UserStatus
from app.models.user import User
from app.services.username import normalize_username


async def ensure_admin_exists(db: AsyncSession) -> None:
    """Mirrors server/src/bootstrap.ts exactly: makes sure the first admin
    account exists on startup. Idempotent — an existing account (including
    its password) is left untouched. Silently no-ops if any of the three
    required env vars is missing, same as the Node version."""
    email = os.environ.get("ADMIN_EMAIL")
    username = os.environ.get("ADMIN_USERNAME")
    password = os.environ.get("ADMIN_PASSWORD")
    if not email or not username or not password:
        return

    username_lower = normalize_username(username)
    result = await db.execute(select(User).where((User.email == email) | (User.usernameLower == username_lower)))
    if result.scalar_one_or_none():
        return

    user = User(
        firstName=os.environ.get("ADMIN_FIRST_NAME") or "Admin",
        lastName=os.environ.get("ADMIN_LAST_NAME") or "Admin",
        email=email,
        username=username,
        usernameLower=username_lower,
        passwordHash=hash_password(password),
        role=Role.ADMIN,
        status=UserStatus.ACTIVE,
        canEditProfile=True,
    )
    db.add(user)
    await db.commit()
    print(f"Created initial admin account: {username}")
